File: scripts/import_xlsx.py
#!/usr/bin/env python3
"""
Bulk-import a model-train inventory spreadsheet into Roundhouse.

Reads an .xlsx with the columns:
    Scale | Mfg | Number | Item | Source | Purchased | Price | Color | Notes

and inserts every row as an item under a single collection. Items are not
assigned to a set; the user can group them in the UI later.

The category prefix on the 'Item' column ("Locomotive - …", "RS - …",
"Track - …", etc.) is mapped to the Roundhouse `type` enum.

Idempotent guard: refuses to run if the target collection already exists.
Pass --force to wipe-and-reimport that collection.

Usage:
    python3 scripts/import_xlsx.py path/to/inventory.xlsx \\
        [--collection NAME] [--description TEXT] [--force]
"""

from __future__ import annotations

import argparse
import sqlite3
import sys
from datetime import datetime, date
from pathlib import Path
from typing import Optional

from openpyxl import load_workbook

# Mac dev path. On Windows this script wouldn't run anyway since Roundhouse
# isn't packaged for Mac use long-term — this is dev-only seed data.
DEFAULT_DB = Path.home() / "Library" / "Application Support" / "roundhouse" / "roundhouse.db"
DEFAULT_COLLECTION_NAME = "Carey's Trains"

VALID_SCALES = {"Z", "N", "HO", "OO", "S", "O", "G"}


def map_scale(raw: object) -> Optional[str]:
    if raw is None:
        return None
    s = str(raw).strip()
    if not s:
        return None
    up = s.upper()
    if up in VALID_SCALES:
        return up
    if up in ("N/A", "ALL", "ANY"):
        return None
    # 1/43, 1/64, ON30, etc — keep but bucket as 'other'
    return "other"


def map_type(item_name: str) -> str:
    if not item_name:
        return "other"
    s = item_name.strip()
    # The category prefix is usually before " - "
    prefix = s.split(" - ", 1)[0].strip().lower()

    if prefix == "rs" or prefix.startswith("rs "):
        return "rolling_stock"
    if prefix.startswith("locomotive"):
        return "locomotive"
    if prefix.startswith("track"):
        return "track"
    if prefix.startswith("accessor"):
        return "accessory"
    if prefix.startswith("building") or "structure" in prefix or "sturcture" in prefix:
        return "building"
    if prefix in ("toy truck", "truck", "ahl", "hartoy"):
        return "accessory"
    if prefix.startswith("transformer") or prefix.startswith("tools"):
        return "accessory"
    if prefix.startswith("trolley"):
        return "rolling_stock"
    # explicit "other" buckets
    if prefix.startswith(
        (
            "puzzle",
            "coin bank",
            "train set",
            "book",
            "repairs",
            "lot",
            "misc",
        )
    ):
        return "other"
    return "other"


def to_iso_date(v: object) -> Optional[str]:
    if v is None:
        return None
    if isinstance(v, datetime):
        return v.date().isoformat()
    if isinstance(v, date):
        return v.isoformat()
    s = str(v).strip()
    return s or None


def to_cents(v: object) -> Optional[int]:
    if v is None:
        return None
    if isinstance(v, (int, float)):
        return int(round(v * 100))
    s = str(v).strip().replace("$", "").replace(",", "")
    if not s:
        return None
    try:
        return int(round(float(s) * 100))
    except ValueError:
        return None


def clean(v: object) -> Optional[str]:
    if v is None:
        return None
    s = str(v).strip()
    if not s or s in ("?", "-"):
        return None
    return s


def main() -> int:
    ap = argparse.ArgumentParser(description="Import a train-inventory .xlsx into Roundhouse")
    ap.add_argument("xlsx", help="Path to the .xlsx inventory file")
    ap.add_argument("--db", default=str(DEFAULT_DB),
                    help="Target SQLite DB (default: Roundhouse user data dir)")
    ap.add_argument("--collection", default=DEFAULT_COLLECTION_NAME,
                    help=f"Collection name to import into (default: {DEFAULT_COLLECTION_NAME!r})")
    ap.add_argument("--description", default=None,
                    help="Collection description (default: 'Imported from <filename>')")
    ap.add_argument("--force", action="store_true",
                    help="Delete the target collection before importing if it exists")
    args = ap.parse_args()

    xlsx_path = Path(args.xlsx).expanduser()
    db_path = Path(args.db).expanduser()
    collection_name: str = args.collection
    collection_desc: str = args.description or f"Imported from {xlsx_path.name}"

    if not xlsx_path.exists():
        print(f"❌ Spreadsheet not found: {xlsx_path}", file=sys.stderr)
        return 2

    if not db_path.exists():
        # Bootstrap the DB by applying the project schema directly.
        schema_path = Path(__file__).resolve().parent.parent / "src" / "main" / "db" / "schema.sql"
        if not schema_path.exists():
            print(f"❌ DB missing and schema.sql not found at {schema_path}", file=sys.stderr)
            return 2
        print(f"🆕 Creating DB at {db_path}")
        db_path.parent.mkdir(parents=True, exist_ok=True)
        boot = sqlite3.connect(db_path)
        boot.executescript(schema_path.read_text())
        boot.execute("PRAGMA journal_mode = WAL")
        boot.commit()
        boot.close()

    print(f"📖 Reading {xlsx_path}")
    wb = load_workbook(xlsx_path, read_only=True, data_only=True)
    ws = wb.active
    rows = list(ws.iter_rows(values_only=True))

    # rows[0] = header, rows[1] = totals row, rows[2+] = data
    # but we tolerate either by filtering for an actual item name.
    data_rows = [r for r in rows[1:] if r[3] and isinstance(r[3], str) and r[3].strip()]
    print(f"   {len(data_rows)} item rows found")

    db = sqlite3.connect(db_path)
    db.execute("PRAGMA foreign_keys = ON")

    # Sanity-check the schema has the source column.
    cols = [r[1] for r in db.execute("PRAGMA table_info(items)").fetchall()]
    if "source" not in cols:
        print("❌ items.source column missing. Wipe the DB and relaunch the app to apply the new schema.",
              file=sys.stderr)
        return 3

    # Sanity-check we have the kind column (v0.5.0+).
    coll_cols = [r[1] for r in db.execute("PRAGMA table_info(collections)").fetchall()]
    has_kind = "kind" in coll_cols
    has_collection_id = "collection_id" in cols

    existing = db.execute("SELECT id FROM collections WHERE name = ?",
                          (collection_name,)).fetchone()
    if existing and not args.force:
        print(f"⚠  Collection '{collection_name}' already exists. Use --force to wipe and reimport.",
              file=sys.stderr)
        return 4
    if existing and args.force:
        print(f"   --force: deleting existing collection (id={existing[0]}) and its sets/items")
        db.execute("DELETE FROM collections WHERE id = ?", (existing[0],))

    if has_kind:
        cur = db.execute(
            "INSERT INTO collections (name, description, kind) VALUES (?, ?, 'trains')",
            (collection_name, collection_desc),
        )
    else:
        cur = db.execute(
            "INSERT INTO collections (name, description) VALUES (?, ?)",
            (collection_name, collection_desc),
        )
    coll_id = cur.lastrowid
    print(f"✓ Created collection '{collection_name}' (id={coll_id})")

    # Counters for the report
    by_type: dict[str, int] = {}
    by_scale: dict[str, int] = {}
    inserted = 0
    skipped = 0

    if has_collection_id:
        insert_sql = """
            INSERT INTO items (
                collection_id, set_id, type, name, manufacturer, model_number, scale,
                road_name, era, year, condition, original_box,
                purchase_date, purchase_price_cents, current_value_cents,
                storage_location, source, notes
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """
    else:
        insert_sql = """
            INSERT INTO items (
                set_id, type, name, manufacturer, model_number, scale,
                road_name, era, year, condition, original_box,
                purchase_date, purchase_price_cents, current_value_cents,
                storage_location, source, notes
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """

    with db:
        for r in data_rows:
            scale_raw, mfg, number, item_name, source, purchased, price, color, notes = r[:9]

            name = clean(item_name)
            if not name:
                skipped += 1
                continue

            type_ = map_type(name)
            scale = map_scale(scale_raw)

            # Color goes into notes if present, since we don't have a color column.
            note_parts = []
            if clean(color):
                note_parts.append(f"Color: {clean(color)}")
            if clean(notes):
                note_parts.append(str(clean(notes)))
            merged_notes = " · ".join(note_parts) if note_parts else None

            base_row = (
                None,                          # set_id
                type_,                         # type
                name,                          # name
                clean(mfg),                    # manufacturer
                clean(number),                 # model_number
                scale,                         # scale
                None,                          # road_name (not in source)
                None,                          # era
                None,                          # year
                None,                          # condition
                None,                          # original_box
                to_iso_date(purchased),        # purchase_date
                to_cents(price),               # purchase_price_cents
                None,                          # current_value_cents
                None,                          # storage_location
                clean(source),                 # source
                merged_notes,                  # notes
            )
            row = (coll_id, *base_row) if has_collection_id else base_row
            db.execute(insert_sql, row)
            inserted += 1
            by_type[type_] = by_type.get(type_, 0) + 1
            if scale:
                by_scale[scale] = by_scale.get(scale, 0) + 1

    print(f"\n✓ Inserted {inserted} item(s); skipped {skipped}")
    print("\n  by type:")
    for t, n in sorted(by_type.items(), key=lambda x: -x[1]):
        print(f"    {t:>14} : {n}")
    print("\n  by scale:")
    for s, n in sorted(by_scale.items(), key=lambda x: -x[1]):
        print(f"    {s:>14} : {n}")

    total = db.execute(
        "SELECT COALESCE(SUM(purchase_price_cents),0) FROM items WHERE id IN "
        "(SELECT id FROM items)"
    ).fetchone()[0]
    print(f"\n  total purchase value: ${total / 100:,.2f}")

    db.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/main/import-xlsx.ts
/**
 * In-app .xlsx importers for trains and coins. Mirrors the column
 * mapping logic from scripts/import_xlsx.py and scripts/import_coins.py
 * so a brand-new user (no Python, no shell) can bring their existing
 * inventory in via the GUI.
 *
 * Each import:
 *   1. Parses the first worksheet of the workbook
 *   2. Walks data rows (skipping headers / totals)
 *   3. Maps each row onto the items insert shape
 *   4. Inserts everything in a single transaction
 *
 * The functions DON'T create the collection — they expect a target
 * collection_id (one already exists per kind from the migration).
 *
 * They return an `ImportResult` so the renderer can show a summary
 * dialog ("Inserted 412 items, skipped 3 blank rows").
 */

import ExcelJS from 'exceljs'
import type Database from 'better-sqlite3'
import { diagLog } from './diag'

export interface ImportResult {
  inserted: number
  skipped: number
  warnings: string[]
}

/**
 * Sanity caps. The Surface user once reported a "small Excel file"
 * importing 1M+ records, almost certainly because the sheet's bounds
 * were silently enormous (a formula or fill-down on a detected column
 * passes the per-row skip check). These caps make that failure mode
 * loud + safe instead of catastrophic.
 */
const MAX_SHEET_ROWS = 100_000   // Refuse upfront — likely a malformed sheet
const MAX_INSERTS = 50_000        // Abort mid-import; the txn rolls back

class ImportLimitError extends Error {
  constructor(message: string) { super(message); this.name = 'ImportLimitError' }
}

/**
 * Cheaply read just the data-row count of the first worksheet without
 * parsing the file twice. Used by the IPC layer to prompt the user
 * before committing to a large import.
 */
export async function peekXlsxRowCount(filePath: string): Promise<number> {
  const wb = new ExcelJS.Workbook()
  await wb.xlsx.readFile(filePath)
  const ws = wb.worksheets[0]
  if (!ws) return 0
  // actualRowCount counts rows with cell values (vs rowCount which is
  // the max touched index). Subtract 1 for the header row.
  return Math.max(0, ws.actualRowCount - 1)
}

function logSheetBounds(ws: ExcelJS.Worksheet, label: string): void {
  // ws.lastRow is the row object for the last used row; .number is its
  // index. actualRowCount counts rows with cell values; rowCount is the
  // max index touched (including blanks). Big mismatches between the
  // two are a tell that the sheet has phantom-formatting rows.
  const last = ws.lastRow?.number ?? 0
  diagLog(
    `[import] sheet bounds (${label}): rowCount=${ws.rowCount} actualRowCount=${ws.actualRowCount} lastRow.number=${last}`
  )
}

// ─── Trains ──────────────────────────────────────────────────────

const VALID_SCALES = new Set(['Z', 'N', 'HO', 'OO', 'S', 'O', 'G'])

function mapScale(raw: unknown): string | null {
  if (raw == null) return null
  const s = String(raw).trim()
  if (!s) return null
  const up = s.toUpperCase()
  if (VALID_SCALES.has(up)) return up
  if (['N/A', 'ALL', 'ANY'].includes(up)) return null
  // 1/43, 1/64, ON30, etc. — keep raw, the Settings page accepts custom scales
  return s
}

function mapTrainType(itemName: string): string {
  if (!itemName) return 'other'
  const prefix = (itemName.split(' - ', 1)[0] ?? '').trim().toLowerCase()
  if (prefix === 'rs' || prefix.startsWith('rs ')) return 'rolling_stock'
  if (prefix.startsWith('locomotive')) return 'locomotive'
  if (prefix.startsWith('track')) return 'track'
  if (prefix.startsWith('accessor')) return 'accessory'
  if (prefix.startsWith('building') || prefix.includes('structure') || prefix.includes('sturcture')) return 'building'
  if (['toy truck', 'truck', 'ahl', 'hartoy'].includes(prefix)) return 'accessory'
  if (prefix.startsWith('transformer') || prefix.startsWith('tools')) return 'accessory'
  if (prefix.startsWith('trolley')) return 'rolling_stock'
  if (
    prefix.startsWith('puzzle') ||
    prefix.startsWith('coin bank') ||
    prefix.startsWith('train set') ||
    prefix.startsWith('book') ||
    prefix.startsWith('repairs') ||
    prefix.startsWith('lot') ||
    prefix.startsWith('misc')
  ) {
    return 'other'
  }
  return 'other'
}

function clean(v: unknown): string | null {
  if (v == null) return null
  // ExcelJS may return rich-text or hyperlink objects; coerce sensibly.
  let s: string
  if (typeof v === 'object' && v !== null) {
    if ('text' in v && typeof (v as { text: unknown }).text === 'string') {
      s = (v as { text: string }).text
    } else if ('result' in v) {
      s = String((v as { result: unknown }).result ?? '')
    } else {
      s = String(v)
    }
  } else {
    s = String(v)
  }
  s = s.trim()
  if (!s || s === '?' || s === '-') return null
  return s
}

function toNumber(v: unknown): number | null {
  if (v == null) return null
  if (typeof v === 'number') return Number.isFinite(v) ? v : null
  const s = String(v).trim().replace(/[$,]/g, '')
  if (!s) return null
  const n = Number(s)
  return Number.isFinite(n) ? n : null
}

function toCents(v: unknown): number | null {
  const n = toNumber(v)
  return n == null ? null : Math.round(n * 100)
}

function toIsoDate(v: unknown): string | null {
  if (v == null) return null
  if (v instanceof Date) {
    return v.toISOString().slice(0, 10)
  }
  const s = String(v).trim()
  return s || null
}

/**
 * Train import. Expected sheet shape (one of the columns near the
 * front; we tolerate variation by name-matching the header row):
 *   Scale | Mfg | Number | Item | Source | Purchased | Price | Color | Notes
 */
export async function importTrainsXlsx(
  db: Database.Database,
  filePath: string,
  collectionId: number
): Promise<ImportResult> {
  const wb = new ExcelJS.Workbook()
  await wb.xlsx.readFile(filePath)
  const ws = wb.worksheets[0]
  if (!ws) {
    return { inserted: 0, skipped: 0, warnings: ['No worksheet found in file.'] }
  }
  logSheetBounds(ws, 'trains')
  if (ws.rowCount > MAX_SHEET_ROWS) {
    return {
      inserted: 0,
      skipped: 0,
      warnings: [
        `This worksheet reports ${ws.rowCount.toLocaleString()} rows — that's almost certainly wrong (the file probably has hidden blank-but-formatted rows extending way past your data). Open the file in Excel, select rows below your last real data row, delete them, save, and try again.`
      ]
    }
  }

  // Detect column indexes from header row. Default fallback is the
  // legacy column order from the Python script.
  const header = ws.getRow(1).values as unknown[]
  const idx = headerIndex(header, {
    scale:     ['scale'],
    mfg:       ['mfg', 'manufacturer'],
    number:    ['number', 'model #', 'model number', 'model_no', 'model'],
    item:      ['item', 'description', 'name'],
    source:    ['source', 'where bought'],
    purchased: ['purchased', 'date', 'purchase date'],
    price:     ['price', 'cost', 'paid'],
    color:     ['color', 'road name', 'road'],
    notes:     ['notes', 'comment', 'comments']
  }, {
    scale: 1, mfg: 2, number: 3, item: 4, source: 5, purchased: 6, price: 7, color: 8, notes: 9
  })

  const insertSql = `
    INSERT INTO items (
      collection_id, type, name,
      manufacturer, model_number, scale, road_name,
      condition, original_box,
      purchase_date, purchase_price_cents,
      source, notes, quantity
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1)
  `
  const insert = db.prepare(insertSql)

  let inserted = 0
  let skipped = 0
  const warnings: string[] = []

  const txn = db.transaction(() => {
    ws.eachRow({ includeEmpty: false }, (row, rowNumber) => {
      if (rowNumber <= 1) return // header (and any phantom row 0)
      if (inserted >= MAX_INSERTS) {
        throw new ImportLimitError(
          `Aborted at ${MAX_INSERTS.toLocaleString()} inserts — the sheet appears to have far more data rows than expected (last row processed: ${rowNumber}). The transaction was rolled back; no items were added. Check the file for a column with fill-down values extending past your real data.`
        )
      }
      const cells = row.values as unknown[]
      const itemRaw = clean(cells[idx.item])
      if (!itemRaw) {
        skipped += 1
        return
      }
      const scale = mapScale(cells[idx.scale])
      const mfg = clean(cells[idx.mfg])
      const number = clean(cells[idx.number])
      const source = clean(cells[idx.source])
      const purchased = toIsoDate(cells[idx.purchased])
      const price = toCents(cells[idx.price])
      const color = clean(cells[idx.color])
      const notes = clean(cells[idx.notes])
      const type = mapTrainType(itemRaw)
      // The "Item" column often carries a "Type - Name" prefix; strip
      // it for the visible name so the Type chip carries that signal.
      const cleanedName = itemRaw.includes(' - ')
        ? itemRaw.split(' - ').slice(1).join(' - ').trim() || itemRaw
        : itemRaw

      insert.run(
        collectionId, type, cleanedName,
        mfg, number, scale, color,
        null, null,
        purchased, price,
        source, notes
      )
      inserted += 1
      if (inserted % 500 === 0) {
        diagLog(`[import] trains progress: ${inserted} inserted at row ${rowNumber}`)
      }
    })
  })

  try {
    txn()
  } catch (err) {
    if (err instanceof ImportLimitError) {
      diagLog(`[import] trains aborted: ${err.message}`)
      return { inserted: 0, skipped: 0, warnings: [err.message] }
    }
    throw err
  }

  if (inserted === 0) {
    warnings.push('No data rows recognized. Make sure the first sheet has an "Item" column with item names.')
  }
  return { inserted, skipped, warnings }
}

// ─── Coins ───────────────────────────────────────────────────────

const COIN_COND_MAP: Record<string, string> = {
  'poor': 'poor', 'p-1': 'poor',
  'fair': 'fair', 'fr-2': 'fair',
  'about good': 'about_good', 'ag-3': 'about_good',
  'good': 'good', 'g-4': 'good', 'g-6': 'good',
  'very good': 'very_good', 'vg-8': 'very_good', 'vg-10': 'very_good',
  'fine': 'fine', 'f-12': 'fine', 'f-15': 'fine',
  'very fine': 'very_fine', 'vf-20': 'very_fine', 'vf-35': 'very_fine',
  'extremely fine': 'extremely_fine', 'ef-40': 'extremely_fine', 'ef-45': 'extremely_fine',
  'about uncirculated': 'about_uncirculated', 'au-50': 'about_uncirculated', 'au-58': 'about_uncirculated',
  'mint': 'mint_state', 'mint state': 'mint_state',
  'uncirculated': 'mint_state', 'unc': 'mint_state',
  'ms-60': 'mint_state', 'ms-65': 'mint_state', 'ms-70': 'mint_state',
  'proof': 'proof', 'pr-60': 'proof', 'pr-70': 'proof'
}

function mapCoinCondition(raw: unknown): string | null {
  const s = clean(raw)
  if (s == null) return null
  return COIN_COND_MAP[s.toLowerCase()] ?? s
}

function mapCoinType(raw: unknown): string {
  const s = (clean(raw) || '').toLowerCase()
  if (s.startsWith('bill')) return 'bill'
  return 'coin'
}

function synthesizeCoinName(
  country: string | null,
  faceValue: number | null,
  denomination: string | null,
  year: number | null,
  mint: string | null
): string {
  const parts: string[] = []
  if (year != null) parts.push(String(year))
  if (country) parts.push(country)
  if (faceValue != null) {
    parts.push(Number.isInteger(faceValue) ? String(faceValue) : String(faceValue))
  }
  if (denomination) parts.push(denomination)
  if (mint) parts.push(`(${mint})`)
  return parts.join(' ') || '(unnamed coin)'
}

/**
 * Coin import. Expected sheet shape:
 *   Type | Country | Currency | Denomination | Year | Mint | Condition |
 *   Quantity | Value | Total | Comment
 *
 * Note: "Currency" in the source spreadsheet holds the *numeric face
 * value*; "Denomination" holds the unit name (Dollar, Pesos…).
 */
export async function importCoinsXlsx(
  db: Database.Database,
  filePath: string,
  collectionId: number
): Promise<ImportResult> {
  const wb = new ExcelJS.Workbook()
  await wb.xlsx.readFile(filePath)
  const ws = wb.worksheets[0]
  if (!ws) {
    return { inserted: 0, skipped: 0, warnings: ['No worksheet found in file.'] }
  }
  logSheetBounds(ws, 'coins')
  if (ws.rowCount > MAX_SHEET_ROWS) {
    return {
      inserted: 0,
      skipped: 0,
      warnings: [
        `This worksheet reports ${ws.rowCount.toLocaleString()} rows — that's almost certainly wrong (the file probably has hidden blank-but-formatted rows extending way past your data). Open the file in Excel, select rows below your last real data row, delete them, save, and try again.`
      ]
    }
  }

  const header = ws.getRow(1).values as unknown[]
  const idx = headerIndex(header, {
    type:         ['type'],
    country:      ['country'],
    currency:     ['currency', 'face value', 'face'],
    denomination: ['denomination', 'denom', 'unit'],
    year:         ['year'],
    mint:         ['mint', 'mint mark'],
    condition:    ['condition', 'grade'],
    quantity:     ['quantity', 'qty', 'count'],
    value:        ['value', 'current value'],
    total:        ['total'],
    comment:      ['comment', 'notes', 'comments']
  }, {
    type: 1, country: 2, currency: 3, denomination: 4, year: 5,
    mint: 6, condition: 7, quantity: 8, value: 9, total: 10, comment: 11
  })

  const insertSql = `
    INSERT INTO items (
      collection_id, type, name,
      country, face_value, denomination, mint_mark, quantity,
      year, condition,
      current_value_cents, notes
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `
  const insert = db.prepare(insertSql)

  let inserted = 0
  let skipped = 0
  const warnings: string[] = []

  const txn = db.transaction(() => {
    ws.eachRow({ includeEmpty: false }, (row, rowNumber) => {
      if (rowNumber <= 1) return // header (and any phantom row 0)
      if (inserted >= MAX_INSERTS) {
        throw new ImportLimitError(
          `Aborted at ${MAX_INSERTS.toLocaleString()} inserts — the sheet appears to have far more data rows than expected (last row processed: ${rowNumber}). The transaction was rolled back; no coins were added. Check the file for a column with fill-down values extending past your real data.`
        )
      }
      const cells = row.values as unknown[]
      const country = clean(cells[idx.country])
      const typeRaw = cells[idx.type]
      // Treat row as data if it has Type or Country populated.
      if (!country && !clean(typeRaw)) {
        skipped += 1
        return
      }
      const t = mapCoinType(typeRaw)
      const faceValue = toNumber(cells[idx.currency])
      const denomination = clean(cells[idx.denomination])
      const yearRaw = toNumber(cells[idx.year])
      const year = yearRaw != null ? Math.trunc(yearRaw) : null
      const mint = clean(cells[idx.mint])
      const condition = mapCoinCondition(cells[idx.condition])
      const qtyRaw = toNumber(cells[idx.quantity])
      const qty = qtyRaw != null ? Math.max(1, Math.trunc(qtyRaw)) : 1
      const valueCents = toCents(cells[idx.value])
      const notes = clean(cells[idx.comment])
      const name = synthesizeCoinName(country, faceValue, denomination, year, mint)

      insert.run(
        collectionId, t, name,
        country, faceValue, denomination, mint, qty,
        year, condition,
        valueCents, notes
      )
      inserted += 1
      if (inserted % 500 === 0) {
        diagLog(`[import] coins progress: ${inserted} inserted at row ${rowNumber}`)
      }
    })
  })

  try {
    txn()
  } catch (err) {
    if (err instanceof ImportLimitError) {
      diagLog(`[import] coins aborted: ${err.message}`)
      return { inserted: 0, skipped: 0, warnings: [err.message] }
    }
    throw err
  }

  if (inserted === 0) {
    warnings.push('No data rows recognized. Make sure the first sheet has Country and/or Type columns populated.')
  }
  return { inserted, skipped, warnings }
}

// ─── Header detection helper ────────────────────────────────────

/**
 * Map a header row into a {column-name → ExcelJS-row-index} record.
 * `aliases[col]` is a list of header text variants (lowercased, trimmed)
 * that map to that column. `defaults` is the fallback 1-based index when
 * no header matches — keeps legacy spreadsheets without proper headers
 * working.
 */
function headerIndex<T extends string>(
  headerRow: unknown[],
  aliases: Record<T, string[]>,
  defaults: Record<T, number>
): Record<T, number> {
  const out = { ...defaults }
  for (let i = 1; i < headerRow.length; i += 1) {
    const cell = headerRow[i]
    if (cell == null) continue
    const normalized = String(cell).trim().toLowerCase()
    if (!normalized) continue
    for (const [col, names] of Object.entries(aliases) as [T, string[]][]) {
      if (names.includes(normalized)) {
        out[col] = i
        break
      }
    }
  }
  return out
}
